iroha_p2p/src/network.rs

```diff
@@ -176,31 +217,69 @@ where
     async fn handle(&mut self, msg: PeerMessage<T>) {
         match msg {
             PeerMessage::Connected(id, connection_id) => {
+                debug!("Connected connection_id {}", connection_id);
                 if self.connected_peers.contains(&id.public_key) {
-                    warn!(
+                    debug!(
                         "Peer with public key {:?} already connected!",
                         &id.public_key
                     );
                     self.broker.issue_send(StopSelf(connection_id)).await;
+                    drop(self.new_peers.remove(&connection_id));
                 } else {
-                    let _ = self.connected_peers.insert(id.public_key);
+                    if !self.new_peers.contains_key(&connection_id) {
+                        sleep(Duration::from_millis(50)).await;
+                    }
+                    if let Some(peer) = self.new_peers.remove(&connection_id) {
+                        drop(self.peers.insert(id.public_key.clone(), peer));
+                        let _ = self.connected_peers.insert(id.public_key);
+                    }
                 }
```

Comment:
This can be made more efficient using [entry api](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry)

Reply:
Where? Can you elaborate?

Reply:
Sorry, Im kinda blind